### PR TITLE
Support querying base benchmark results when commenting pull requests

### PR DESCRIPTION
**Pull request by antiagainst**
Directly show how the performance of a pull request looks like
against the current base branch's tip. This avoids the author
to jump hops to get the numbers so less friction. ;)

### CODE REVIEW

**Comment by antiagainst** [comment]
We will be looking at something like the following on pull requests after this change (ignoring all the numbers and links; they are mock):

### Benchmark results

@ commit e1b4f5f62aa3b1dd9076420d3a0684cb64f479f3 (vs. base 3e067747b4c98f479f8b9ea52040b77265d3fd1a)

| Benchmark Name                                                                                                        | Average Latency (ms)  | Median Latency (ms) | Latency Standard Deviation (ms) |
| --------------------------------------------------------------------------------------------------------------------- | --------------------- | ------------------- | ------------------------------- |
| MobileBertSquad (TensorFlow) fp32-full-inference-1-thread with IREE-Dylib-Sync @ SM-G980F (CPU-ARMv8.2-A)             | 778 (vs. 777, 0.13%↑) | 776                 | 9                               |
| MobileBertSquad (TensorFlow) fp32-full-inference-3-thread with IREE-Dylib @ SM-G980F (CPU-ARMv8.2-A)                  | 759 (vs. 769, 1.30%↓) | 724                 | 223                             |
| MobileNetV2 [imagenet] (TensorFlow) fp32-full-inference with IREE-Vulkan @ SM-G980F (GPU-Mali-G77)                    | 41 (vs. 41, 0.00%)    | 41                  | 1                               |
| MobileNetV2 [imagenet] (TensorFlow) fp32-full-inference-1-thread with IREE-Dylib-Sync @ SM-G980F (CPU-ARMv8.2-A)      | 79 (vs. 79, 0.00%)    | 79                  | 1                               |
| MobileNetV2 [imagenet] (TensorFlow) fp32-full-inference-3-thread with IREE-Dylib @ SM-G980F (CPU-ARMv8.2-A)           | 222 (vs. 222, 0.00%)  | 224                 | 30                              |
| MobileNetV2 [imagenet] (TensorFlow) fp32-kernel-execution with IREE-Vulkan @ SM-G980F (GPU-Mali-G77)                  | 23 (vs. 23, 0.00%)    | 23                  | 2                               |
| MobileNetV3Small [imagenet] (TensorFlow) fp32-full-inference with IREE-Vulkan @ SM-G980F (GPU-Mali-G77)               | 45 (vs. 45, 0.00%)    | 45                  | 4                               |
| MobileNetV3Small [imagenet] (TensorFlow) fp32-full-inference-1-thread with IREE-Dylib-Sync @ SM-G980F (CPU-ARMv8.2-A) | 34 (vs. 34, 0.00%)    | 34                  | 0                               |
| MobileNetV3Small [imagenet] (TensorFlow) fp32-full-inference-3-thread with IREE-Dylib @ SM-G980F (CPU-ARMv8.2-A)      | 188 (vs. 188, 0.00%)  | 186                 | 9                               |
| MobileNetV3Small [imagenet] (TensorFlow) fp32-kernel-execution with IREE-Vulkan @ SM-G980F (GPU-Mali-G77)             | 24 (vs. 24, 0.00%)    | 25                  | 1                               |

See more details on [Buildkite](https://buildkite.com/iree/iree-android-benchmark/builds/2575)

**Comment by benvanik** [comment]
as a note, dylib-sync and dylib with 1 thread are completely different - you'll want dylib for the 1-thread. if you want sync benchmarks then you'll want a separate type for that

**Comment by antiagainst** [comment]
> as a note, dylib-sync and dylib with 1 thread are completely different - you'll want dylib for the 1-thread. if you want sync benchmarks then you'll want a separate type for that

Yup. I learnt that from Ahmed too.  In the above Dylib and Dylib-Sync are treated as separate IREE drivers; so adding more configurations for either of them (including 1 thread for Dylib) is very trivial. I will save that for a dedicated pull request as example for how to registering new benchmarks. :)

**Comment by benvanik** [comment]
👍 
mostly pointing it out because the naming you have there is weird:
image

**Comment by antiagainst** [comment]
> 👍
> mostly pointing it out because the naming you have there is weird:
> image

Oh these are just showing what we have right now in the existing benchmark pipeline. I think in its fullness for Dylib/Dylib-Sync we will have:

```
MobileNetV2 [fp32,imagenet] (TensorFlow) 1-thread,big-core,full-inference with IREE-Dylib-Sync @ SM-G980F (CPU-ARMv8.2-A)
MobileNetV2 [fp32,imagenet] (TensorFlow) 1-thread,little-core,full-inference with IREE-Dylib-Sync @ SM-G980F (CPU-ARMv8.2-A)

MobileNetV2 [fp32,imagenet] (TensorFlow) 1-thread,big-core,full-inference with IREE-Dylib @ SM-G980F (CPU-ARMv8.2-A)
MobileNetV2 [fp32,imagenet] (TensorFlow) 1-thread,little-core,full-inference with IREE-Dylib @ SM-G980F (CPU-ARMv8.2-A)
MobileNetV2 [fp32,imagenet] (TensorFlow) 3-thread,big-core,full-inference with IREE-Dylib @ SM-G980F (CPU-ARMv8.2-A)
MobileNetV2 [fp32,imagenet] (TensorFlow) 3-thread,little-core,full-inference with IREE-Dylib @ SM-G980F (CPU-ARMv8.2-A)
```

Or maybe we can drop the `1-thread` for dylib-sync, as I think it's always 1 thread?

File: build_tools/android/post_benchmarks_as_pr_comment.py [hunk 1]
```diff
@@ -95,24 +135,69 @@ def aggregate_all_benchmarks(
...
+    if pr > base:
+      percent = "{:.2%}".format((pr - base) / base)
+      direction = "↑"
+    elif pr == base:
+      percent = "{:.2%}".format((pr - base) / base)
+      direction = ""
+    else:
+      percent = "{:.2%}".format((base - pr) / base)
+      direction = "↓"
+    comparisions.append(f"{pr} (vs. {base}, {percent}{direction})")
```
Comment (ScottTodd) [inline build_tools/android/post_benchmarks_as_pr_comment.py:167]: Could we emphasize the `↑` / `↓` to make these easier to scan?

---

Current appearance:

| Benchmark Name                                                                                                        | Average Latency (ms)  | Median Latency (ms) | Latency Standard Deviation (ms) |
| --------------------------------------------------------------------------------------------------------------------- | --------------------- | ------------------- | ------------------------------- |
| MobileBertSquad (TensorFlow) fp32-full-inference-1-thread with IREE-Dylib-Sync @ SM-G980F (CPU-ARMv8.2-A)             | 778 (vs. 777, 0.13%↑) | 776                 | 9                               |

---

Maybe use emojis or images?

Time increased: ⬆️🔺🙁
Time decreased: ⬇️🔻🙂

Or colors? https://stackoverflow.com/questions/11509830/how-to-add-color-to-githubs-readme-md-file (multiple options)

**Review by ScottTodd — APPROVED** [review]

File: build_tools/android/post_benchmarks_as_pr_comment.py (hunk 1 above)
Comment (antiagainst) [reply build_tools/android/post_benchmarks_as_pr_comment.py]: Excellent points!

Using colors would be the best; it's disappointing that GitHub still does not support it.

I'm using 🚩 and 🎉 for now if the percentage change is larger than a threshold. :)

**Pull request merged**